test/Expression/SubExpression/JoinCommonSubExpressions.purs

```diff
@@ -0,0 +1,86 @@
+module Test.Expression.SubExpression.JoinCommonSubExpressions
+  ( joinCommonSubExpressionsTests
+  ) where
+
+import Prelude
+
+import Data.Either (Either(..))
+import Expression.Error (Expect, throw)
+import Expression.Parser (parse)
+import Expression.SubExpression (joinCommonSubExpressions)
+import Expression.Syntax (Expression)
+import Test.Unit (TestSuite, suite, test)
+import Test.Unit.Assert (equal)
+
+joinCommonSubExpressionsTests :: TestSuite
+joinCommonSubExpressionsTests =
+  suite "Expression.SubExpression - joinCommonSubExpressions" do
+    test "ASSERT f(x) = let $v1 = sinx in let $v2 = sin$v1 in let $v3 = sin$v2 in let $v4 = $v3+$v2 in $v4+$v1 WHEN f(x) = sin(sin(sin(x))) + sin(sin(x)) + sin(x)" do
+      let
+        -- given
+        rawExpression = "sin(sin(sin(x))) + sin(sin(x)) + sin(x)"
+
+        -- when
+        result = fromExpect $ parseAndJoinCommonSubExpressions rawExpression
+
+        -- then
+        expectedResult = "let $v1 = sinx in let $v2 = sin$v1 in let $v3 = sin$v2 in let $v4 = $v3+$v2 in $v4+$v1"
+      equal expectedResult result
+    test "ASSERT f(x) = let $v1 = sinx in let $v2 = sin$v1 in $v2+$v1 WHEN f(x) = sin(sin(x)) + sin(x)" do
+      let
+        -- given
+        rawExpression = "sin(sin(x)) + sin(x)"
+
+        -- when
+        result = fromExpect $ parseAndJoinCommonSubExpressions rawExpression
+
+        -- then
+        expectedResult = "let $v1 = sinx in let $v2 = sin$v1 in $v2+$v1"
+      equal expectedResult result
+    test "ASSERT f(x) = x WHEN f(x) = x" do
+      let
+        -- given
+        rawExpression = "x"
+
+        -- when
+        result = fromExpect $ parseAndJoinCommonSubExpressions rawExpression
+
+        -- then
+        expectedResult = "x"
+      equal expectedResult result
+    test "ASSERT f(x) = x+x WHEN f(x) = x+x" do
+      let
+        -- given
+        rawExpression = "x+x"
+
+        -- when
+        result = fromExpect $ parseAndJoinCommonSubExpressions rawExpression
+
+        -- then
+        expectedResult = "x+x"
+      equal expectedResult result
+    test "ASSERT f(x) = let $v1 = x+x in $v1+$v1 WHEN f(x) = (x+x)+(x+x)" do
+      let
+        -- given
+        rawExpression = "(x+x)+(x+x)"
+
+        -- when
+        result = fromExpect $ parseAndJoinCommonSubExpressions rawExpression
+
+        -- then
+        expectedResult = "let $v1 = x+x in $v1+$v1"
+      equal expectedResult result
+
```

Comment:
Good tests.  To be safer, I think we should add also f(x) = sin((x+x) + (x+x)) + (x+x) + sin((x+x) + (x+x)).

Reply:
I added the test you suggested and the output is correct just the variable numbers are not in the correct order.
`let $v2 = x+x in let $v4 = $v2+$v2 in let $v1 = sin$v4 in let $v3 = $v1+$v2 in $v3+$v1`